cfdm/data/abstract/mesharray.py

```diff
@@ -0,0 +1,366 @@
+
+        :Returns:
+
+            `int`
+                The position of the cell dimension.
+
+        """
+        return self._get_component("cell_dimension", default)
+
+    def get_start_index(self, default=ValueError()):
+        """Return the start index.
+
+        .. versionadded:: (cfdm) UGRIDVER
+
+        :Parameters:
+
+            default: optional
+                Return the value of the *default* parameter if there
+                is no start index.
+
+                {{default Exception}}
+
+        :Returns:
+
+            `int`
+                The start index.
+
+        """
+        return self._get_component("start_index", default)
+
+    def subarray_parameters(self):
+        """Non-data parameters required by the `Subarray` class.
+
+        .. versionadded:: (cfdm) UGRIDVER
+
+        :Returns:
+
+            `dict`
+                The parameters, with keys ``'compressed_dimensions'``,
+                ``'start_index'``, and ``'cell_dimension'``.
+
+        """
+        return {
+            **super().subarray_parameters(),
+            "start_index": self.get_start_index(),
+            "cell_dimension": self.get_cell_dimension(),
+        }
+
+    def subarray_shapes(self, shapes):
+        """Create the subarray shapes along each uncompressed dimension.
+
+        Note that the ouput is indpendent of the *shapes* parameter,
+        because each dimension of the compressed data corresponds to a
+        unique dimension of the uncompressed data.
+
+        .. versionadded:: (cfdm) UGRIDVER
+
+        .. seealso:: `subarray`
+
+        :Parameters:
+
+            {{subarray_shapes chunks: `int`, sequence, `dict`, or `str`, optional}}
+
+        :Returns:
+
+            `list`
+                The subarray sizes along each uncompressed dimension.
+
+        **Examples**
+
+        >>> a.shape
+        (4, 4)
+        >>> a.compressed_dimensions()
+        {0: (0,), 1: (1,)}
+        >>> a.subarray_shapes(-1)
+        [(4,), (4)]
+        >>> a.subarray_shapes("auto")
+        [(4,), (4)]
+        >>> a.subarray_shapes("60B")
+        [(4,), (4)]
+
+        """
+        if shapes == -1:
+            return [(size,) for size in self.shape]
+
+        u_dims = self.get_compressed_axes()
+
+        if isinstance(shapes, (str, Number)):
+            return [
+                (size,) if i in u_dims else shapes
+                for i, size in enumerate(self.shape)
+            ]
+
+        if isinstance(shapes, dict):
+            shapes = [
+                shapes[i] if i in shapes else None for i in range(self.ndim)
+            ]
+        elif len(shapes) != self.ndim:
+            raise ValueError(
+                f"Wrong number of 'shapes' elements in {shapes}: "
+                f"Got {len(shapes)}, expected {self.ndim}"
+            )
+
+        # chunks is a sequence
+        return [
+            (size,) if i in u_dims else c
+            for i, (size, c) in enumerate(zip(self.shape, shapes))
+        ]
```

Comment:
Elegant implementation here - a very functional (as in 'functional programming') method given all the list comprehensions.